Change Tensor/TensorImpl to use c10::intrusive_ptr

ezyang opened the pull request:
Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/10824

API additions:
- Tensor(c10::intrusive_ptr<TensorImpl,UndefinedTensor>&&)
- Tensor(const c10::intrusive_ptr<TensorImpl,UndefinedTensor>&)
- Tensor::operator=(Tensor&&) && (for completeness sake)
- TensorBase::unsafeGetTensorImpl()
- TensorBase::unsafeReleaseTensorImpl()
- TensorBase::getIntrusivePtr()
- TensorImpl::type_id()
- Tensor::set_data()
- Tensor::is_same(Tensor)
- Tensor::use_count()
- Tensor::type_id()
- Tensor::scalar_type()
- WeakTensor::is_same(WeakTensor)
- intrusive_ptr::weak_use_count()
- weak_intrusive_ptr::weak_use_count()
- c10::raw::intrusive_ptr::{incref,decref,make_weak}
- c10::raw::weak_intrusive_ptr::{incref,decref,lock}

API changes:
- Tensor::pImpl is no longer public (and now named tensor_impl_)
    - Most methods accessed this way are now accessible on Tensor
      maybe_zero_dim() and set_wrapped_number() being prominent exceptions
      (they are now accessed through unsafeGetTensorImpl())
- Type is no longer friend of Tensor
- TensorBase::reset(TensorImpl*) is deleted
- TensorBase::reset(TensorImpl*, bool should_retain) is deleted
- TensorBase::swap(TensorBaseImpl&) is deleted; use std::swap instead
- TensorBase::get() is deleted; use unsafeGetTensorImpl() instead
- TensorBase::detach() is deleted; use unsafeReleaseTensorImpl() instead
- TensorBase::retain() is deleted; use _raw_incref() instead
- TensorBase::release() is deleted; use _raw_decref() instead
- WeakTensor lost most of its methods (it no longer inherits from
  TensorBase)
- TensorImpl::storage() is now a const method
- Tensor(TensorBase) constructor removed, instead
  we go through getIntrusivePtr().  I'm not sure about
  this change; I happened to have accidentally removed the
  TensorBase constructor and decided to fix call sites,
  but I could go the other way.
- detail::set_data() is deleted; use Tensor::set_data() instead
- c10::raw_intrusive_ptr_target removed; use the functions in c10::raw instead.
  (The reason for this change, is that it is invalid to cast an intrusive_ptr_target*
  to a raw_intrusive_ptr_target* to take advantage of the methods. But there is
  no reason the incref/decref methods shouldn't also work on intrusive_ptr_target;
  it is primarily an API consideration. We can be more standards compliant by
  keeping them as functions, which are universally applicable.)
- intrusive_ptr::reclaim() and weak_intrusive_ptr::reclaim() now work on
  pointers of the NullType. (This counts as a bug fix, because the documentation
  specified that pointers produced by release() are valid to reclaim(), and
  a release() on a null intrusive_ptr produces the NullType::singleton())

Bug fixes:
- Dispatch code for mutable references incorrectly returned
  a reference to a value argument (which would immediately
  go out of scope).  They now correctly return a tensor by
  value.
- intrusive_ptr copy/move assignment did not work correctly when
  an object was assigned to itself. We now check for this case and
  no-op if so. (This bug manifested itself as a Tensor mysteriously
  becoming an UndefinedTensor after lines of code like
  'x = x.mul_(y)'

Other changes:
- The checked cast functions in Utils.h have now been
  renamed and detemplatized into checked unwrap functions.
- Added type_id() and scalar_type() methods to Tensor
- pImpl is no longer public
- Documented what the && overloads are doing
- All occurrences of 'new TensorImpl' (and similar spellings, like 'new THTensor')
  have been expunged. This is NO LONGER a valid way to create a new
  tensor, and if you do this, upon your first incref, you will catch an ASSERT
  failure saying that only tensors created by intrusive_ptr::release() are valid
  to reclaim(). Use c10::make_intrusive instead in this situation.
- IValue is adjusted to use intrusive_ptr instead of Retainable, and all
  other sub-classes of Retainable were modified to use intrusive_ptr.
  When doing this, I had to make the constructors of sub-classes like
  ConstantList public, so that c10::make_intrusive could invoke them.  Fortunately,
  if you incorrectly stack allocate a ConstantList, and then try to get an
  intrusive_ptr to it, it will fail, as stack allocated ConstantLists have refcount 0.
- IValue very narrowly sidesteps the problem of handling NullType, as it
  considers intrusive_ptr<TensorImpl> identical to intrusive_ptr<TensorImpl, UndefinedTensor>
  which is not always true. This was always the case, but there's now a comment
  explaining what's going on.

Some MSVC bugs were uncovered during the preparation of this patch.
They are documented as comments in the code.

Differential Revision: D9481140

goldsborough commented on torch/csrc/api/include/torch/nn/cloneable.h:
```diff
@@ -58,9 +58,7 @@ class Cloneable : public virtual Module {
-        at::detail::set_data(
-            copy->parameters_[parameter.key],
-            autograd::Variable(*parameter).data().clone());
+        copy->parameters_[parameter.key].set_data(autograd::Variable(*parameter).data().clone());
```
This also changes the variable type, like `detail::set_data` used to? Otherwise LGTM, although `set_data` is more "public" than `at::detail::set_data` was intended to be (!)

ezyang replied:
I'm fine with prefixing it with an underscore to make it more private. The primary matter at hand was I didn't want to go through the trouble of friending this function.

gchanan commented on aten/src/ATen/templates/TypeDerived.cpp:
```diff
@@ -91,11 +91,7 @@ Storage ${Type}::storageWithAllocator(int64_t size, Allocator* allocator) const
-    pimpl->retain();
-  }
-  return Tensor(pimpl, false);
+  return Tensor(static_cast<TensorImpl*>(th_pointer), retain);
```
this will retain an UndefinedTensor where it didn't before -- do we want that?

ezyang replied:
It's OK, because the `intrusive_ptr` codepath no-ops in this case. (It's a little hard to see, but we incref by reclaiming a pointer, and then making a copy of the `intrusive_ptr`. Reclaiming UndefinedTensor* used to be illegal but I relaxed the assert in `intrusive_ptr`, and copies of `intrusive_ptr`, when an `UndefinedTensor` is pointed at, are just no-ops, as you would expect.)

gchanan commented on aten/src/ATen/templates/TensorMethods.h:
```diff
@@ -9,10 +9,6 @@
 
 namespace at {
 
-inline Tensor & Tensor::operator=(Tensor const & rhs) && {
```
you need this right?

ezyang replied:
Yes I do! :)

gchanan commented on torch/csrc/jit/ivalue.h:
```diff
@@ -161,7 +161,7 @@ struct IValue {
   : tag(Tag::Tensor), retainable(t.defined())  {
     // note: the undefined tensor is not refcounted, so while it
     // is tagged as a tensor, retainable is set to false.
-    as_tensor_impl = t.at::detail::TensorBase::detach();
```
I feel like I'm missing something.  Why are you mentioning `detach` in python?  I mentioned the semantics change because "t.TensorBase::detach() meant "return the raw pointer, relinquishing ownership on it." and now you aren't doing that.

ezyang replied:
Oh, I think your confusion is about what `unsafeReleaseTensorImpl` does. This function is NOT a decref (actually, there is no way to decref the smart pointer class Tensor, and rightly so). It is, "relinquish control of the TensorImpl* pointer, and return it.) I'm using `release()` and `get()` nomenclature consistent with how `unique_ptr` names them.

gchanan commented on aten/src/ATen/UndefinedTensor.h:
```diff
@@ -6,15 +6,15 @@ namespace at {
 struct AT_API UndefinedTensor final : public TensorImpl {
 public:
-  static inline UndefinedTensor * singleton() {
+  static constexpr inline TensorImpl * singleton() {
```
hmm...are you sure you want to use UndefinedTensor in this way?  Did you do it because it makes writing a few functions easier in TensorBase?

ezyang replied:
I'm not sure what you mean by this question. I guess I can try restating the justification: if you `s/TensorImpl/UndefinedTensor/`, a static assert in `c10::intrusive_ptr` fails, because when you pass UndefinedTensor as the NullType to `c10::intrusive_ptr<TensorImpl, UndefinedTensor>`, it checks that `UndefinedTensor::singleton()` returns a `TensorImpl*` (NOT the refined type.)

As it turns out, no one really needs the refined `UndefinedTensor` type (because it's purpose in life is to serve as an implementation of the `TensorImpl` interface), so this change didn't have any other consequences.

gchanan replied:
Sorry for not being clear here: what I meant was "are you sure you want to use UndefinedTensor as the NullType singleton", not "are you sure you want to return something of type UndefinedTensor."  We discussed the former and decided to do it to avoid increfs/decrefs.

ezyang commented on aten/src/ATen/TensorBase.h:
```diff
@@ -2,107 +2,52 @@
-using TensorBase = TensorBaseImpl<true>;
-using WeakTensorBase = TensorBaseImpl<false>;
+protected:
+  c10::intrusive_ptr<TensorImpl, UndefinedTensor> tensor_impl_;
```
Two reasons: (1) Storage prior to this patch was using `storage_impl_` so I wanted to make it consistent, and (2) pImpl was historically used to refer to `TensorImpl*` raw pointer, so when I changed the type I wanted force myself to update all of the call sites (in case something typechecked, but was wrong).

ezyang commented:
I've successfully squashed the last MSVC bug. The summary has been updated with all the new changes.

If people would like to see smaller patches for review, I can split it up (some of the changes are easy to split). But I am lazy, so I am defaulting to not doing this for now.

ezyang committed:
Change Tensor/TensorImpl to use c10::intrusive_ptr (#10824)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/10824

API additions:
- Tensor(c10::intrusive_ptr<TensorImpl,UndefinedTensor>&&)
- Tensor(const c10::intrusive_ptr<TensorImpl,UndefinedTensor>&)
- Tensor::operator=(Tensor&&) && (for completeness sake)
- TensorBase::unsafeGetTensorImpl()
- TensorBase::unsafeReleaseTensorImpl()
- TensorBase::getIntrusivePtr()
- TensorImpl::type_id()
- Tensor::set_data()
- Tensor::is_same(Tensor)
- Tensor::use_count()
- Tensor::type_id()
- Tensor::scalar_type()
- WeakTensor::is_same(WeakTensor)
- intrusive_ptr::weak_use_count()
- weak_intrusive_ptr::weak_use_count()
- c10::raw::intrusive_ptr::{incref,decref,make_weak}
- c10::raw::weak_intrusive_ptr::{incref,decref,lock}

API changes:
- Tensor::pImpl is no longer public (and now named tensor_impl_)
    - Most methods accessed this way are now accessible on Tensor
      maybe_zero_dim() and set_wrapped_number() being prominent exceptions
      (they are now accessed through unsafeGetTensorImpl())
- Type is no longer friend of Tensor
- TensorBase::reset(TensorImpl*) is deleted
- TensorBase::reset(TensorImpl*, bool should_retain) is deleted
- TensorBase::swap(TensorBaseImpl&) is deleted; use std::swap instead
- TensorBase::get() is deleted; use unsafeGetTensorImpl() instead
- TensorBase::detach() is deleted; use unsafeReleaseTensorImpl() instead
- TensorBase::retain() is deleted; use _raw_incref() instead
- TensorBase::release() is deleted; use _raw_decref() instead
- WeakTensor lost most of its methods (it no longer inherits from
  TensorBase)
- TensorImpl::storage() is now a const method
- Tensor(TensorBase) constructor removed, instead
  we go through getIntrusivePtr().  I'm not sure about
  this change; I happened to have accidentally removed the
  TensorBase constructor and decided to fix call sites,
  but I could go the other way.
- detail::set_data() is deleted; use Tensor::set_data() instead
- c10::raw_intrusive_ptr_target removed; use the functions in c10::raw instead.
  (The reason for this change, is that it is invalid to cast an intrusive_ptr_target*
  to a raw_intrusive_ptr_target* to take advantage of the methods. But there is
  no reason the incref/decref methods shouldn't also work on intrusive_ptr_target;
  it is primarily an API consideration. We can be more standards compliant by
  keeping them as functions, which are universally applicable.)
- intrusive_ptr::reclaim() and weak_intrusive_ptr::reclaim() now work on
  pointers of the NullType. (This counts as a bug fix, because the documentation
  specified that pointers produced by release() are valid to reclaim(), and
  a release() on a null intrusive_ptr produces the NullType::singleton())

Bug fixes:
- Dispatch code for mutable references incorrectly returned
  a reference to a value argument (which would immediately
  go out of scope).  They now correctly return a tensor by
  value.
- intrusive_ptr copy/move assignment did not work correctly when
  an object was assigned to itself. We now check for this case and
  no-op if so. (This bug manifested itself as a Tensor mysteriously
  becoming an UndefinedTensor after lines of code like
  'x = x.mul_(y)')

Other changes:
- The checked cast functions in Utils.h have now been
  renamed and detemplatized into checked unwrap functions.
- Added type_id() and scalar_type() methods to Tensor
- pImpl is no longer public
- Documented what the && overloads are doing
- All occurrences of 'new TensorImpl' (and similar spellings, like 'new THTensor')
  have been expunged. This is NO LONGER a valid way to create a new
  tensor, and if you do this, upon your first incref, you will catch an ASSERT
  failure saying that only tensors created by intrusive_ptr::release() are valid
  to reclaim(). Use c10::make_intrusive instead in this situation.
- IValue is adjusted to use intrusive_ptr instead of Retainable, and all
  other sub-classes of Retainable were modified to use intrusive_ptr.
  When doing this, I had to make the constructors of sub-classes like
  ConstantList public, so that c10::make_intrusive could invoke them.  Fortunately,
  if you incorrectly stack allocate a ConstantList, and then try to get an
  intrusive_ptr to it, it will fail, as stack allocated ConstantLists have refcount 0.
- IValue very narrowly sidesteps the problem of handling NullType, as it
  considers intrusive_ptr<TensorImpl> identical to intrusive_ptr<TensorImpl, UndefinedTensor>
  which is not always true. This was always the case, but there's now a comment
  explaining what's going on.

Some MSVC bugs were uncovered during the preparation of this patch.
They are documented as comments in the code.

Differential Revision: D9481140

fbshipit-source-id: 67dc3f28a1bf68863d1b8e2e5d73696fe00fe139

gchanan commented on tools/autograd/gen_python_functions.py:
```diff
@@ -447,6 +447,26 @@ def append_actuals_formals(actual, formal):
+        #
+        #   Tensor& dispatch_selu_(Tensor self) { return at::selu_(self); }
+        #
+        # (NB: We can't make dispatch_selu_ take Tensor&, because the enclosing
```
I think the code you want here is already supported if you pass `unpack_args=True` to `parse_arg`.  Might be better just to do that in this case (we already do it if there is a tensoroptions for example).  But this is probably fine too.
